### ISSUE TITLE
False positive with 'string|object|callable'

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#257]
```php
<?php
/**
 * @param string|object|callable $param
 */
function f($param): void {}
f('BOOM!');
```
results in 'Argument 1 of f expects object, string provided' even though it accepts string

**Comment by weirdan** [comment]
Of all the possible permutations of param types there's a single one that does not exhibit this problem: `callable|string|object` (e() signature below)
```php
<?php
/**
 * @param string|object|callable $param
 */
function a($param): void {}
/**
 * @param object|string|callable $param
 */
function b($param): void {}
/**
 * @param object|callable|string $param
 */
function c($param): void {}
/**
 * @param callable|object|string $param
 */
function d($param): void {}
/**
 * @param callable|string|object $param
 */
function e($param): void {}
/**
 * @param string|callable|object $param
 */
function f($param): void {}

a('BOOM!');
b('BOOM!');
c('BOOM!');
d('BOOM!');
e('BOOM!');
f('BOOM!');
```

**Comment by muglug** [comment]
Thanks for opening this!

**Comment by muglug** [comment]
And all those above cases are fixed by that

**Comment by weirdan** [comment]
With a34de5b all of the above passes successfully.
